### ISSUE TITLE
[clang-tidy] Check request: detect saving stack addresses beyond their lifetime

### ISSUE THREAD

**Issue by asund** [llvm/llvm-project#122985]
This seems to be missed by existing stack address check as the address doesn't escape scope of the stack but is preserved between calls using a static variable.
```
auto f() {
  process stack_array[] = { method1, method2, method3 };
  static *process process_to_use = nullptr;

  if (!process) {
    // some expensive init later...
    process = &stack_array[n];
  }

  if (!process) {
   process->do_processing();  // segfault
  }
}
```
process_to_use has a stale value when the function is called again. stack_array needs to have static lifetime in this case.

**Comment by EugeneZelenko** [comment]
I thin Clang Static Analyzer is better place for such check.

**Comment by chrchr-github** [comment]
Should it say `process_to_use->do_processing();` in the example?

**Comment by chrchr-github** [comment]
This compilable example is caught by `clang-analyzer-core.StackAddressEscape`:
~~~c++
void g() {}
using fp_t = void (*)();
void f() {
  fp_t stack_array[] = { g };
  static fp_t* fp = nullptr;

  if (!fp) {
    fp = &stack_array[0];
  }

  (*fp)();
}

int main() {
    f();
    f();
}
~~~
https://godbolt.org/z/oscdKv5Mz
